File: 0064-minimum-path-sum/0064-minimum-path-sum.py
class Solution(object):
    def minPathSum(self, grid):
        ROW, COL = len(grid), len(grid[0])

        res = [[float("inf")] * (COL + 1) for i in range(ROW + 1)]
        res[ROW][COL - 1] = 0

        for i in range(ROW-1, -1, -1):
            for j in range(COL-1, -1, -1):
                res[i][j] = grid[i][j] + min(res[i+1][j], res[i][j+1])

        return res[0][0]

File: 0064-minimum-path-sum/0064-minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int rows = grid.size(), cols = grid[0].size();
        vector<vector<int>> dp(rows, vector<int>(cols, 0));

        dp[0][0] = grid[0][0];
        for (int i = 1; i < cols; i++) dp[0][i] = grid[0][i] + dp[0][i-1];
        for (int i = 1; i < rows; i++) dp[i][0] = grid[i][0] + dp[i-1][0];

        for (int row = 1; row < rows; row++){
            for (int col = 1; col < cols; col++){
                dp[row][col] = grid[row][col] + min(dp[row - 1][col], dp[row][col - 1]);
            }
        }

        return dp[rows - 1][cols - 1];
    }
};